pkg/accountmanager/manager.go

```diff
@@ -257,3 +294,152 @@ func (a *AccountManager) getAccountPoller(name *types.NamespacedName) (*accountP
+	if acctConfig == nil {
+		return fmt.Errorf("failed to get account config")
+	}
+	config, ok := acctConfig.selectorConfigMap[*selectorNamespacedName]
+	if !ok {
+		config = &selectorConfig{
+			namespacedName: selectorNamespacedName,
+			selector:       selector,
+		}
+		a.Log.V(1).Info("Adding selector config", "account",
+			accountNamespacedName, "selector", selectorNamespacedName)
+		acctConfig.selectorConfigMap[*selectorNamespacedName] = config
+		return nil
+	}
+
+	// Update the selector with the latest copy.
+	a.Log.V(1).Info("Updating selector config", "account",
+		accountNamespacedName, "selector", selectorNamespacedName)
+	config.selector = selector
+	return nil
+}
+
+// removeSelectorConfig removes the current selector configuration for the specified account.
+func (a *AccountManager) removeSelectorConfig(accountNamespacedName, selectorNamespacedName *types.NamespacedName) {
+	acctConfig := a.getAccountConfig(accountNamespacedName)
+	if acctConfig == nil {
+		a.Log.Error(fmt.Errorf("failed to get account config"), "", "account", accountNamespacedName)
+		return
+	}
+	a.Log.V(1).Info("Removing selector config", "account", accountNamespacedName, "selector", selectorNamespacedName)
+	delete(acctConfig.selectorConfigMap, *selectorNamespacedName)
+}
+
+func (a *AccountManager) getAccountProviderType(namespacedName *types.NamespacedName) (common.ProviderType, bool) {
+	acctConfig := a.getAccountConfig(namespacedName)
+	if acctConfig != nil {
+		return acctConfig.providerType, true
+	}
+
+	return "", false
+}
+
+// handleAddProviderAccountError performs the cleanup on account add/update.
+// i.e. Removes poller, inventory and returns if the error can be retried or not.
+func (a *AccountManager) handleAddProviderAccountError(namespacedName *types.NamespacedName, config *accountConfig, err error) bool {
+	// Account poller is removed upon any error in the plug-in.
+	if err := a.removeAccountPoller(namespacedName); err != nil {
+		a.Log.Error(err, "error removing account poller")
+	}
+	_ = a.Inventory.DeleteVpcsFromCache(namespacedName)
+	_ = a.Inventory.DeleteVmsFromCache(namespacedName)
+	// TODO: require lock to write into account config structure.
+	config.accountInitState = false
+	if strings.Contains(err.Error(), util.ErrorMsgSecretReference) {
+		config.credentialsValid = false
+		config.retry = false
+	} else {
+		config.credentialsValid = true
+		config.retry = true
+	}
+	return config.retry
+}
+
+// replaySelectorsForAccount replays all the selectors that belong to this specified account,
+// set/resets the selector status.
+func (a *AccountManager) replaySelectorsForAccount(namespacedName *types.NamespacedName, config *accountConfig) {
+	for _, filterConfig := range config.selectorConfigMap {
+		a.Log.Info("Re-playing selector", "account", namespacedName, "selector", filterConfig.namespacedName)
+		_, err := a.AddResourceFiltersToAccount(namespacedName, filterConfig.namespacedName, filterConfig.selector, true)
```

Comment:
update filter directly by calling plugin.

Reply:
We would also need to update the selectors in the poller. Doing so, we will end having another version of the AddResourceFiltersToAccount() function. So avoiding it.